### PR TITLE
auto_buffer: replace usage with std::vector

### PR DESCRIPTION
**Pull request by lrusak**
This removes the legacy `XFILE::auto_buffer` implementation and uses `std::vector` instead.

`auto_buffer` uses realloc which is unnecessary in c++. Here is some reading about this claim:
https://www.stroustrup.com/bs_faq2.html#realloc
https://tinodidriksen.com/2012/02/cpp-vector-vs-realloc/
I compiled the test file on my system and found this
```
$ ./vector-realloc
realloc() reallocations: 1587
std::vector reallocations: 24
```
This is purely based on the fact that `std::vector` has a growth factor and doesn't grab the free'd block after the `realloc`. YMMV.

I used `std::vector<char>` as `auto_buffer` used `char` for the underlying interpretation. I could use `uint8_t` in some cases but I think it's ok (unless someone has strong feelings).

### CODE REVIEW

File: xbmc/NfoFile.cpp [hunk 1]
```diff
@@ -131,10 +131,10 @@ int CNfoFile::Load(const std::string& strFile)
 {
   Close();
   XFILE::CFile file;
-  XFILE::auto_buffer buf;
+  std::vector<char> buf;
```
Comment (bugwelle) [inline xbmc/NfoFile.cpp:134]: Hi, just a question from an interested user: Why not use `std::string` directly as the buffer? :)

At least in this file, it seems that actual characters are stored and not binary data (It's about NFO files after all).

Thanks.

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecTX3G.cpp [hunk 2]
```diff
@@ -190,8 +186,8 @@ int CDVDOverlayCodecTX3G::Decode(DemuxPacket* pPacket)
         if (curRecord.endChar > textLength)
           curRecord.endChar = textLength;
 
-        bgnStyle.get()[curRecord.bgnChar] |= curRecord.faceStyleFlags;
-        endStyle.get()[curRecord.endChar] |= curRecord.faceStyleFlags;
+        bgnStyle.data()[curRecord.bgnChar] |= curRecord.faceStyleFlags;
```
Comment (bugwelle) [inline xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecTX3G.cpp:189]:
```suggestion
        bgnStyle[curRecord.bgnChar] |= curRecord.faceStyleFlags;
```

Why the extra `.data()`? (possibly due to search&replace?)

**Comment by arnova** [comment]
Looks good but I would use uint8_t instead of char when used for bytes/binary stuff. Furthermore you can drop the reinterpret_cast()s then.

File: xbmc/filesystem/File.cpp [hunk 3]
```diff
@@ -815,9 +815,9 @@ ssize_t CFile::Write(const void* lpBuf, size_t uiBufSize)
     { // "test" write with zero size
       // some VFSs don't handle correctly null buffer pointer
       // provide valid buffer pointer for them
-      auto_buffer dummyBuf(255);
-      dummyBuf.get()[0] = 0;
-      return m_pFile->Write(dummyBuf.get(), 0);
+      std::vector<char> dummyBuf(255);
+      dummyBuf.data()[0] = 0;
+      return m_pFile->Write(dummyBuf.data(), 0);
```
Comment (neo1973) [inline xbmc/filesystem/File.cpp:820]: I would change this to
```
const char dummyData = 0;
return m_pFile->Write(&dummyData, 0);
```
No need to construct a `std::vector` and to allocate on the heap.

File: xbmc/platform/win32/filesystem/Win32File.cpp [hunk 4]
```diff
@@ -238,10 +237,10 @@ ssize_t CWin32File::Write(const void* lpBuf, size_t uiBufSize)
 
   if (uiBufSize == 0)
   { // allow "test" write with zero size
-    XUTILS::auto_buffer dummyBuf(255);
-    dummyBuf.get()[0] = 0;
+    std::vector<char> dummyBuf(255);
+    dummyBuf.data()[0] = 0;
     DWORD bytesWritten = 0;
-    if (!WriteFile(m_hFile, dummyBuf.get(), 0, &bytesWritten, NULL))
+    if (!WriteFile(m_hFile, dummyBuf.data(), 0, &bytesWritten, NULL))
```
Comment (neo1973) [inline xbmc/platform/win32/filesystem/Win32File.cpp:243]: Like above I suggest to change this to
```
const char dummyData = 0;
DWORD bytesWritten = 0;
if (!WriteFile(&dummyData, 0, &bytesWritten, NULL))
```

File: xbmc/platform/win32/filesystem/Win32File.cpp [hunk 5]
```diff
@@ -577,8 +576,9 @@ int CWin32File::Stat(const CURL& url, struct __stat64* statData)
     if (lastDot != std::wstring::npos && pathnameW.rfind(L'\\') < lastDot)
     { // file has some extension
       const std::wstring fileExt(pathnameW, lastDot);
-      XUTILS::auto_buffer buf(32767 * sizeof(wchar_t)); // maximum possible size
-      const DWORD envRes = GetEnvironmentVariableW(L"PATHEXT", (wchar_t*)buf.get(), buf.size() / sizeof(wchar_t));
+      std::vector<char> buf(32767 * sizeof(wchar_t)); // maximum possible size
```
Comment (neo1973) [inline xbmc/platform/win32/filesystem/Win32File.cpp:579]: Change to `std::vector<wchar_t>` and get rid of the casts and the extra size calculation?

File: xbmc/NfoFile.cpp (hunk 1 above)
Comment (lrusak) [reply xbmc/NfoFile.cpp]: I've updated to use `std::string` I'm not sure if I like it or not yet as it seems a bit awkward in the usage of `CFile::LoadFile`.

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecTX3G.cpp (hunk 2 above)
Comment (lrusak) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecTX3G.cpp]: updated, thanks.

File: xbmc/filesystem/File.cpp (hunk 3 above)
Comment (lrusak) [reply xbmc/filesystem/File.cpp]: updated

File: xbmc/platform/win32/filesystem/Win32File.cpp (hunk 4 above)
Comment (lrusak) [reply xbmc/platform/win32/filesystem/Win32File.cpp]: updated

File: xbmc/platform/win32/filesystem/Win32File.cpp (hunk 5 above)
Comment (lrusak) [reply xbmc/platform/win32/filesystem/Win32File.cpp]: done

File: xbmc/NfoFile.cpp (hunk 1 above)
Comment (bugwelle) [reply xbmc/NfoFile.cpp]: I don't mind any of both ways. I was just asking myself what the benefit of one vs the other was. :)

Comment (neo1973) [reply xbmc/NfoFile.cpp]: Using `std::string` in the context of `CFile::LoadFile` seems very wrong to me. When loading arbitrary files who says there aren't `'\0'` in the middle and who guarantees there is a `'\0'` at the end?

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp [hunk 6]
```diff
@@ -95,16 +95,17 @@ bool CDVDSubtitleStream::Open(const std::string& strFile)
...
   if (!pInputStream)
     return true;
 
   static const uint8_t vobsub[] = { 0x00, 0x00, 0x01, 0xBA };
 
-  int read =
-      pInputStream->Read(reinterpret_cast<uint8_t*>(buf.data()), static_cast<int>(buf.size()));
+  //! @todo: fix with c++17
+  int read = pInputStream->Read(const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(buf.data())),
+                                static_cast<int>(buf.size()));
```
Comment (neo1973) [inline xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp:108]: Like in my other comment, who guarantees that there will be a `'\0'` at the end? Thinking also about corrupt files here... `std::vector` looks like the correct way to me.

Comment (lrusak) [reply xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp]: I've updated it to use `std::vector`

File: xbmc/guilib/Texture.cpp [hunk 7]
```diff
@@ -286,9 +285,7 @@ bool CTexture::LoadFromFileInternal(const std::string& texturePath,
     pImage = ImageFactory::CreateLoaderFromMimeType(strMimeType);
 
   //! @todo: c++17 should be able to fix this casting situation
```
Comment (garbear) [inline xbmc/guilib/Texture.cpp:287]: Comment can be removed

**Review by garbear — APPROVED** [review]

File: xbmc/platform/win32/filesystem/Win32File.cpp [hunk 8]
```diff
@@ -177,9 +176,9 @@ ssize_t CWin32File::Read(void* lpBuf, size_t uiBufSize)
 
   if (uiBufSize == 0)
   { // allow "test" read with zero size
-    XUTILS::auto_buffer dummyBuf(255);
+    std::vector<char> dummyBuf(255);
```
Comment (neo1973) [inline xbmc/platform/win32/filesystem/Win32File.cpp:179]: A `const char dummyBuf = 0;` would be sufficient.

**Review by neo1973 — APPROVED** [review]

File: xbmc/guilib/Texture.cpp (hunk 7 above)
Comment (lrusak) [reply xbmc/guilib/Texture.cpp]: thanks

File: xbmc/platform/win32/filesystem/Win32File.cpp (hunk 8 above)
Comment (lrusak) [reply xbmc/platform/win32/filesystem/Win32File.cpp]: not sure how i missed this one, thanks

**Review by garbear — APPROVED** [review]

**Comment by lrusak** [comment]
rebased and squashed, could probably use another once over.

**Review by garbear — APPROVED** [review]
I think I saw a place for some move optimization, but the current state of the PR is isomorphic to the original code as far as memcpys go so no complaints from me.

**Comment by lrusak** [comment]
jenkins build this please

**Pull request merged**